File: src/main/java/io/github/shaksternano/mediamanipulator/emoji/EmojiUtil.java
package io.github.shaksternano.mediamanipulator.emoji;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.google.gson.*;
import io.github.shaksternano.mediamanipulator.Main;
import io.github.shaksternano.mediamanipulator.io.FileUtil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

public class EmojiUtil {

    public static final String EMOJI_FILES_DIRECTORY = "src/main/resources/" + FileUtil.getResourcePathInRootPackage("emoji");
    private static Set<String> emojiUnicodeSet = ImmutableSet.of();
    private static Map<String, String> emojiShortcodesToUrls = ImmutableMap.of();

    public static void initEmojiUnicodeSet() {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(FileUtil.getResourceInRootPackage("emoji/emoji_unicodes.txt")))) {
            ImmutableSet.Builder<String> builder = ImmutableSet.builder();
            String line;
            while ((line = reader.readLine()) != null) {
                builder.add(line.toLowerCase());
            }
            emojiUnicodeSet = builder.build();
        } catch (IOException e) {
            Main.getLogger().error("Error while loading emoji unicodes!", e);
        }
    }

    public static void initEmojiShortCodesToUrlsMap() {
        try (Reader reader = new InputStreamReader(FileUtil.getResourceInRootPackage("emoji/emojis.json"))) {
            JsonElement emojiNamesElement = JsonParser.parseReader(reader);
            if (emojiNamesElement instanceof JsonObject emojiNamesObject) {
                emojiShortcodesToUrls = emojiNamesObject.entrySet()
                    .stream()
                    .filter(entry -> {
                        boolean isString = entry.getValue() instanceof JsonPrimitive primitive && primitive.isString();
                        if (!isString) {
                            Main.getLogger().error("Error parsing shortcodes file! The value {} of the entry with key {} is not a string!", entry.getValue(), entry.getKey());
                        }
                        return isString;
                    })
                    .map(entry -> {
                        String emoji = entry.getValue().getAsString();
                        String emojiUnicode = emoji.codePoints()
                            .mapToObj(Integer::toHexString)
                            .collect(Collectors.joining("-"));
                        return Map.entry(entry.getKey(), getEmojiUrl(emojiUnicode));
                    })
                    .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
            } else {
                Main.getLogger().error("Invalid emoji shortcodes JSON:\n{}", emojiNamesElement);
            }
        } catch (IOException e) {
            Main.getLogger().error("Error while loading emoji shortcodes!", e);
        } catch (JsonParseException e) {
            Main.getLogger().error("Error while parsing emoji shortcodes!", e);
        }
    }

    public static boolean isEmojiUnicode(String unicode) {
        return emojiUnicodeSet.contains(unicode.toLowerCase());
    }

    public static Optional<String> getEmojiUrlFromShortcode(String shortcode) {
        return Optional.ofNullable(emojiShortcodesToUrls.get(shortcode));
    }

    public static String getEmojiUrl(String unicode) {
        return "https://raw.githubusercontent.com/twitter/twemoji/master/assets/72x72/" + unicode + ".png";
    }
}

File: core/src/main/kotlin/emoji/EmojiUtil.kt
package com.shakster.borgar.core.emoji

import com.shakster.borgar.core.io.IO_DISPATCHER
import com.shakster.borgar.core.io.getResource
import com.shakster.borgar.core.logger
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.stream.Collectors

const val EMOJI_FILES_DIRECTORY: String = "core/src/main/resources/emoji"

private val emojiUnicodeSet: MutableSet<String> = mutableSetOf()
private val emojiShortcodesToUrls: MutableMap<String, String> = mutableMapOf()

suspend fun initEmojis() {
    initEmojiUnicodeSet()
    initEmojiShortCodesToUrlsMap()
}

private suspend fun initEmojiUnicodeSet() {
    runCatching {
        val lines = withContext(IO_DISPATCHER) {
            BufferedReader(InputStreamReader(getResource("emoji/emoji_unicodes.txt"))).use {
                it.readLines()
            }
        }
        emojiUnicodeSet.clear()
        val lowercaseLines = lines.map { it.lowercase() }
        emojiUnicodeSet.addAll(lowercaseLines)
    }.onFailure {
        logger.error("Error while loading emoji unicodes!", it)
    }
}

private suspend fun initEmojiShortCodesToUrlsMap() {
    runCatching {
        val jsonString = withContext(IO_DISPATCHER) {
            BufferedReader(InputStreamReader(getResource("emoji/emojis.json"))).use {
                it.readText()
            }
        }
        val emojis = Json.decodeFromString<Map<String, String>>(jsonString)
        val emojiUrls = emojis.mapValues { (_, emoji) ->
            val emojiUnicode = emoji.codePoints()
                .mapToObj(Integer::toHexString)
                .collect(Collectors.joining("-"))
            getEmojiUrl(emojiUnicode)
        }
        emojiShortcodesToUrls.clear()
        emojiShortcodesToUrls.putAll(emojiUrls)
    }.onFailure {
        logger.error("Error while loading emoji shortcodes", it)
    }
}

fun isEmojiUnicode(unicode: String): Boolean {
    return emojiUnicodeSet.contains(unicode.lowercase())
}

fun getEmojiUrlFromShortcode(shortcode: String): String? =
    emojiShortcodesToUrls[shortcode]

fun getEmojiUrl(unicode: String): String =
    "https://raw.githubusercontent.com/twitter/twemoji/master/assets/72x72/$unicode.png"
